src/provider.c

```diff
@@ -1218,74 +1248,84 @@ int OSSL_provider_init(const OSSL_CORE_HANDLE *handle, const OSSL_DISPATCH *in,
+            p11prov_cfg_names[i].name, (char **)&cfg[i], sizeof(void *));
+    }
+    core_params[P11PROV_CFG_SIZE] = OSSL_PARAM_construct_end();
+
     ret = core_get_params(handle, core_params);
     if (ret != RET_OSSL_OK) {
         ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_GET_PARAMETER);
         p11prov_ctx_free(ctx);
         return ret;
     }
 
-    ret = p11prov_module_new(ctx, path, init_args, &ctx->module);
+    ret = p11prov_module_new(ctx, cfg[P11PROV_CFG_PATH],
+                             cfg[P11PROV_CFG_INIT_ARGS], &ctx->module);
     if (ret != CKR_OK) {
         ERR_raise(ERR_LIB_PROV, PROV_R_IN_ERROR_STATE);
         p11prov_ctx_free(ctx);
         return RET_OSSL_ERR;
     }
 
-    if (pin != NULL) {
-        ret = p11prov_get_pin(pin, &ctx->pin);
+    if (cfg[P11PROV_CFG_TOKEN_PIN] != NULL) {
+        ret = p11prov_get_pin(cfg[P11PROV_CFG_TOKEN_PIN], &ctx->pin);
         if (ret != 0) {
             ERR_raise(ERR_LIB_PROV, PROV_R_IN_ERROR_STATE);
             p11prov_ctx_free(ctx);
             return RET_OSSL_ERR;
         }
     }
 
-    if (allow_export != NULL) {
+    if (cfg[P11PROV_CFG_ALLOW_EXPORT] != NULL) {
         char *end = NULL;
         errno = 0;
-        ctx->allow_export = (int)strtol(allow_export, &end, 0);
+        ctx->allow_export = (int)strtol(cfg[P11PROV_CFG_ALLOW_EXPORT], &end, 0);
         if (errno != 0 || *end != '\0') {
             P11PROV_raise(ctx, CKR_GENERAL_ERROR, "Invalid value for %s: (%s)",
-                          P11PROV_PKCS11_MODULE_ALLOW_EXPORT, allow_export);
+                          p11prov_cfg_names[P11PROV_CFG_ALLOW_EXPORT],
+                          cfg[P11PROV_CFG_ALLOW_EXPORT]);
             p11prov_ctx_free(ctx);
             return RET_OSSL_ERR;
         }
     }
 
-    if (login_behavior != NULL) {
-        if (strcmp(login_behavior, "auto") == 0) {
+    if (cfg[P11PROV_CFG_LOGIN_BEHAVIOR] != NULL) {
+        if (strcmp(cfg[P11PROV_CFG_LOGIN_BEHAVIOR], "auto") == 0) {
             ctx->login_behavior = PUBKEY_LOGIN_AUTO;
-        } else if (strcmp(login_behavior, "always") == 0) {
+        } else if (strcmp(cfg[P11PROV_CFG_LOGIN_BEHAVIOR], "always") == 0) {
             ctx->login_behavior = PUBKEY_LOGIN_ALWAYS;
-        } else if (strcmp(login_behavior, "never") == 0) {
+        } else if (strcmp(cfg[P11PROV_CFG_LOGIN_BEHAVIOR], "never") == 0) {
             ctx->login_behavior = PUBKEY_LOGIN_NEVER;
         } else {
             P11PROV_raise(ctx, CKR_GENERAL_ERROR, "Invalid value for %s: (%s)",
-                          P11PROV_PKCS11_MODULE_LOGIN_BEHAVIOR, login_behavior);
+                          p11prov_cfg_names[P11PROV_CFG_LOGIN_BEHAVIOR],
+                          cfg[P11PROV_CFG_LOGIN_BEHAVIOR]);
             p11prov_ctx_free(ctx);
             return RET_OSSL_ERR;
         }
     }
 
-    if (load != NULL && strcmp(load, "early") == 0) {
+    if (cfg[P11PROV_CFG_CACHE_PINS] != NULL
+        && strcmp(cfg[P11PROV_CFG_CACHE_PINS], "cache") == 0) {
```

Comment:
I do not want to use a boolean in the config because I think we may see more than 2 options. the code is boolean because that is an internal detail and we can change it later at will.